doc/api18970.md

```diff
@@ -0,0 +1,334 @@
+
+For API behavior we are primarily interested in whether invoking a
+function can cause a context switch from a cooperative thread.
+
+### Context Terminology
+
+**TODO** fill this out
+
+This section is intended to describe the privilege and processor context
+variations in which a thread can run.  It should provide or reference
+definitions of terms like this:
+
+* Kernel [initialization
+  level](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels)
+* User-space versus kernel (system call)
+* Normal ("thread"?) versus interrupt (invoked from an Interrupt Service
+  Routine)
+
+For API behavior we are interested in whether a particular function
+**may**, **must**, or **must not** be invoked from a specific context.
+
+### Function vs Operation
+
+tl;dr: A function *returns* while an operation *completes*.
+
+A **function** is an addressible sequence of actions to which control is
+transferred by **invoking** the function with various parameters, and
+from which control is transferred back to the point where it was invoked
+when the function **returns**.
+
+A function may succeed or fail.  In many cases failure is indicated by a
+negative integer return value while success is indicated by a
+non-negative integer return value, but specific functions may use other
+conventions.
+
+An **operation** is behavior, which is generally initiated by invoking a
+function.  An operation can also succeed or fail.  An operation
+**completes** when the success or failure of the operation has been
+determined and made available through whatever mechanism was specified
+through the initiating function.
+
+For APIs we are particularly interested in the relationship between the
+function and its associated operation.  Some potential relationships
+include:
+* The function succeeds if and only if its associated operation
+  completes with success. (This is usually expected of *synchronous*
+  functions.)
+* If the function fails then its associated operation has not been
+  initiated.  If the function succeeds then its associated operation has
+  been initiated.  The success or failure of the operation is not
+  otherwise coupled to success or failure of its initiating function.
+  (This is usually expected of *asynchronous* functions.)
+
+## Definitions
+
+### rescheduling (function)
+
+A function is rescheduling when it invokes an operation that includes a
+reschedule point (potentially, but not necessarily, context switching to
+a non-current thread).
+
+Note that whether a reschedule-point function will cause a context
+switch depends on the priority of the current and ready threads.
+
+#### Commentary
+
+The term *rescheduling* applies only to functions that explicitly invoke
+operations that (may) encounter a reschedule point.  Unless interrupts
+are disabled any function may be context-switched if the interrupt
+causes a meta-irq thread to become ready.
+
+### reschedule-safe (function)
```

Comment:
Did you consider calling this "cooperative-safe", and if so, why did you decide against it?

Reply:
I did not consider that, but it seems to work especially when paired with cooperative-unsafe.